validation/charts/appco/istio_test.go

```diff
@@ -0,0 +1,220 @@
+	istioChart, logCmd, err := installIstioAppCo(client, i.cluster.ID, *AppCoUsername, *AppCoAccessToken, gatewayNamespaceSets)
+	require.NoError(i.T(), err)
+	require.True(i.T(), strings.Contains(logCmd, expectedDeployLog))
+	require.True(i.T(), istioChart.IsAlreadyInstalled)
+}
+
+func (i *IstioTestSuite) TestCanaryUpgrade() {
+	subSession := i.session.NewSession()
+	defer subSession.Cleanup()
+
+	client, err := i.client.WithSession(subSession)
+	require.NoError(i.T(), err)
+
+	i.T().Logf("Creating %s namespace", charts.RancherIstioNamespace)
+	err = createIstioNamespace(client, i.cluster.ID)
+	require.NoError(i.T(), err)
+
+	i.T().Logf("Creating %s secret", RancherIstioSecret)
+	logCmd, err := createIstioSecret(client, i.cluster.ID, *AppCoUsername, *AppCoAccessToken)
+	require.NoError(i.T(), err)
+	require.True(i.T(), strings.Contains(logCmd, RancherIstioSecret))
+
+	i.T().Log("Installing SideCar Istio AppCo")
+	istioChart, logCmd, err := installIstioAppCo(client, i.cluster.ID, *AppCoUsername, *AppCoAccessToken, "")
+	require.NoError(i.T(), err)
+	require.True(i.T(), strings.Contains(logCmd, expectedDeployLog))
+	require.True(i.T(), istioChart.IsAlreadyInstalled)
+
+	i.T().Log("Running Canary Istio AppCo Upgrade")
+	istioChart, logCmd, err = upgradeIstioAppCo(client, i.cluster.ID, *AppCoUsername, *AppCoAccessToken, "--set istiod.revision=canary,base.defaultRevision=canary")
+	require.NoError(i.T(), err)
+	require.True(i.T(), strings.Contains(logCmd, expectedDeployLog))
+	require.True(i.T(), istioChart.IsAlreadyInstalled)
+
+	i.T().Log("Verifying if istio-ingress gateway is using the canary revision")
+	getCanaryCommand := []string{
+		"sh", "-c",
+		fmt.Sprintf(`kubectl -n %s get pod -o jsonpath='{.items..metadata.name}'`, charts.RancherIstioNamespace),
+	}
+
+	appName := canaryRevisionApp
+	logCmd, err = kubectl.Command(client, nil, i.cluster.ID, getCanaryCommand, "2MB")
+	require.NoError(i.T(), err)
+	require.True(i.T(), strings.Contains(logCmd, appName))
+}
+
+func (i *IstioTestSuite) TestInPlaceUpgrade() {
+	subSession := i.session.NewSession()
+	defer subSession.Cleanup()
+
+	client, err := i.client.WithSession(subSession)
+	require.NoError(i.T(), err)
+
+	i.T().Logf("Creating %s namespace", charts.RancherIstioNamespace)
+	err = createIstioNamespace(client, i.cluster.ID)
+	require.NoError(i.T(), err)
+
+	i.T().Logf("Creating %s secret", RancherIstioSecret)
+	logCmd, err := createIstioSecret(client, i.cluster.ID, *AppCoUsername, *AppCoAccessToken)
+	require.NoError(i.T(), err)
+	require.True(i.T(), strings.Contains(logCmd, RancherIstioSecret))
+
+	i.T().Log("Installing SideCar Istio AppCo")
+	istioChart, logCmd, err := installIstioAppCo(client, i.cluster.ID, *AppCoUsername, *AppCoAccessToken, "")
+	require.NoError(i.T(), err)
+	require.True(i.T(), strings.Contains(logCmd, expectedDeployLog))
+	require.True(i.T(), istioChart.IsAlreadyInstalled)
+
+	i.T().Log("Running In Place Istio AppCo Upgrade")
```

Comment:
what makes this 'in place' vs. a normal upgrade?

Reply:
hmm, @rancher-max thoughts on using in-place vs. leaving it out? putting it in makes it sound like a special case imo.

Reply:
It does make it sound like a special case, but I think it's fine to leave it in since that's how it is described in the documentation. Ref: https://docs.apps.rancher.io/reference-guides/istio/#upgrade-the-chart